crates/topos-p2p/src/behaviour/gossip.rs

```diff
@@ -0,0 +1,282 @@
+                        echos.data.push(data);
+                    } else {
+                        break;
+                    }
+                }
+
+                P2P_GOSSIP_BATCH_SIZE.observe(echos.data.len() as f64);
+                let msg = bincode::serialize::<Batch>(&echos).expect("msg ser");
+
+                _ = self.gossipsub.publish(IdentTopic::new(TOPOS_ECHO), msg);
+            }
+
+            if !self.ready_queue.is_empty() {
+                let mut readies = Batch { data: Vec::new() };
+                for _ in 0..self.batch_size {
+                    if let Some(data) = self.ready_queue.pop_front() {
+                        readies.data.push(data);
+                    } else {
+                        break;
+                    }
+                }
+
+                P2P_GOSSIP_BATCH_SIZE.observe(readies.data.len() as f64);
+                let msg = bincode::serialize::<Batch>(&readies).expect("msg ser");
+
+                _ = self.gossipsub.publish(IdentTopic::new(TOPOS_READY), msg);
+            }
+        }
+
+        let event = match self.gossipsub.poll(cx, params) {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(ToSwarm::GenerateEvent(event)) => event,
+            Poll::Ready(ToSwarm::ListenOn { opts }) => {
+                return Poll::Ready(ToSwarm::ListenOn { opts })
+            }
+            Poll::Ready(ToSwarm::RemoveListener { id }) => {
+                return Poll::Ready(ToSwarm::RemoveListener { id })
+            }
+            Poll::Ready(ToSwarm::Dial { opts }) => return Poll::Ready(ToSwarm::Dial { opts }),
+            Poll::Ready(ToSwarm::NotifyHandler {
+                peer_id,
+                handler,
+                event,
+            }) => {
+                return Poll::Ready(ToSwarm::NotifyHandler {
+                    peer_id,
+                    handler,
+                    event,
+                })
+            }
+            Poll::Ready(ToSwarm::CloseConnection {
+                peer_id,
+                connection,
+            }) => {
+                return Poll::Ready(ToSwarm::CloseConnection {
+                    peer_id,
+                    connection,
+                })
+            }
+            Poll::Ready(ToSwarm::ExternalAddrExpired(addr)) => {
+                return Poll::Ready(ToSwarm::ExternalAddrExpired(addr))
+            }
+            Poll::Ready(ToSwarm::ExternalAddrConfirmed(addr)) => {
+                return Poll::Ready(ToSwarm::ExternalAddrConfirmed(addr))
+            }
+            Poll::Ready(ToSwarm::NewExternalAddrCandidate(addr)) => {
+                return Poll::Ready(ToSwarm::NewExternalAddrCandidate(addr))
+            }
+        };
+
+        if let gossipsub::Event::Message { ref message_id, .. } = event {
+            if self.cache.contains(message_id) {
+                P2P_DUPLICATE_MESSAGE_ID_RECEIVED.inc();
+            }
+        }
+
+        let outcome = match event {
```

Comment:
Could simplify to something like that?

```rust
let outcome = match event {
    gossipsub::Event::Message {
        propagation_source,
        message_id,
        message: Message {
            source,
            data,
            sequence_number,
            topic,
        },
    } => {
        if [TOPOS_GOSSIP, TOPOS_ECHO, TOPOS_READY].contains(&topic.as_str()) {
            return Poll::Ready(ToSwarm::GenerateEvent(ComposedEvent::Gossipsub(
                crate::event::GossipEvent {
                    topic,
                    message: data,
                    source,
                },
            )));
        }
    }
    _ => {}
};

```

Reply:
mmmh not really as `GossipEvent` uses a `&'static str` as `topic`. I replace the `match event` by a `if let` tho